docs/rfc/rfc-1.adoc

```diff
@@ -0,0 +1,143 @@
+assets to maintain the peg. Depositors send BTC funds to the
+most-recently-created-wallet by using pay-to-script-hash (P2SH) which contains
+hashed information about the depositor's minting ethereum address. Once
+the block is mined, the depositor reveals their desired ethereum minting
+address to the ethereum chain. The TBTC system listens for these sorts of
+messages and when it gets one, it checks the bitcoin network to make sure the
+funds line up. If everything checks out, then we mint!
+
+=== Redemption
+
+A user supplies tBTCv2 and a bitcoin address. The system burns the tBTCv2 and
+releases the equivalent amount of bitcoin to the user from the oldest
+outstanding bitcoin wallet. If the oldest does not have enough, the remaining
+balance carries over to the next oldest until an equivalent amount of BTC has
+been released.
+
+=== Continuous Fees
+
+The system is able to collect fees in the form of tBTC when the BTC is
+deposited and when it is redeemed. Rather than paying operators at those two
+instances we would prefer to pay them continuously. Since we do not have the
+tBTC to pay them with until we collect the redemption fee (and can't mint it
+and maintain the peg), we must turn to another token. Instead, the system pays
+operators continuously in mintable T tokens and whenever it collects tBTC fees
+it uses those fees to buy T tokens back for its treasury.
+
+== In Depth
+
+=== Minting
+
+==== Deposit + Commitment
+
+Once we know the active wallet's public key hash, the dApp can put together a
+pay-to-script-hash (P2SH) address to receive the funds. This script will be
+unique to each depositor and will look like:
+
+```
+<eth-address> DROP
+<blinding-factor> DROP
+DUP HASH160 <signingGroupPubkey> EQUAL
+IF
+  CHECKSIG
+ELSE
+  DUP HASH160 <refundPubkey> EQUALVERIFY
+  <locktime> CHECKLOCKTIMEVERIFY DROP
+  CHECKSIG
+ENDIF
+```
+
+Since each depositor has their own ethereum address and their own secret
+blinding factor (which is an additional security layer), each depositor's
+script will be unique, and the hash of each depositor's script will be unique.
+
+In order to unlock the funds, one must provide the unhashed script, (which
+means that they know the eth address and blinding factor), as well as an
+unlocking script with a signature and public key. If the sig+pubkey matches the
+signing group public key, the funds are able to be moved immediately. If the
+sig+pubkey matches the refund public key, then the funds can be moved after 30
+days (specified as `locktime`).
+
+==== The Big Reveal
+
+After the deposit transaction has been mined, the user is able to reveal their
+ethereum address and blinding factor to the ethereum chain. The TBTC system listens
+for these sorts of messages and when it sees one, is able to generate a script that
+can spend the funds. Once successful, wallet moves the funds from the P2SH
+address to the main wallet address, which allows the system to mint tBTC and
+disables the 30-day refund.
```

Comment:
When is TBTC minted? I think we agreed on minting it immediately once the script is revealed and then sweeping BTC periodically. For the first iteration, would it be enough to sweep immediately?

Reply:
Minting before sweeping is risky. If an attacker gains the ability to DoS a wallet, they can deposit a large amount, minting the TBTC (and selling it before the rest of the attack is revealed), deliberately fail to sweep it, and thus take the 30-day refund while also getting the same BTC on-chain.

Reply:
Consider the following:

We have wallets A and B. Each holds 100 BTC and there are 200 TBTC outstanding. Mallory starts with 10 BTC and is able to selectively DoS wallet B.

Wallet A: 100 BTC
Wallet B: 100 BTC

Mallory: 10 BTC

Mallory deposits 10 BTC in wallet B, receiving 10 TBTC. Mallory redeems 10 TBTC. The situation is now:

Wallet A: 100 BTC (-10 awaiting redemption to Mallory)
Wallet B: 100 BTC (+10 awaiting sweep)

Mallory: 0 BTC (+10 pending redemption, +10 refundable if not swept)

Mallory prevents wallet B from sweeping the deposit, but wallet A completes Mallory's redemption:

Wallet A: 90 BTC
Wallet B: 100 BTC (+10 awaiting sweep)

Mallory: 10 BTC (+10 refundable)

Mallory can now repeat the attack by re-depositing the BTC and redeeming it again. Once wallet A is empty, Mallory either stops the attack, with 10 BTC at hand and 90 BTC refundable in 30 days, or keeps depositing BTC and lets wallet B complete the redemption transactions but not the deposit sweeps. Even if other members of wallet B don't generate redemption transactions before the pending sweeps complete, Mallory is able to loot all BTC held by non-targeted wallets with an almost arbitrarily small initial capital investment.

Reply:
Ah, and this is would be thwarted by sweeping before minting. 

Then, Mallory won't get 10 TBTC until Wallet B has 110 BTC. She's able to effectively move funds to a wallet that she can't DoS to a wallet that she can by paying deposit/redemption fees, but isn't able to de-peg.

Reply:
Yep. If depositing gives you a NFT representing the UTXO and sweeping mints to the holder of the NFT, we can keep the peg safe and provide a quickminting service which redeems unswept UTXO NFTs for TBTC without the holders having to wait for the sweep. Then if someone manages to DoS the sweeps they can only drain the quickminting service instead of the entirety of TBTC.

Reply:
[https://github.com/keep-network/tbtc-research/pull/3/files#diff-93ad61cca4bd3006ea9d4948930a3e30b95a52de58ae02d0246e034ef85a6f1b](Here's) my earlier prototype for the quickminting. Because ethereum nonces enforce ordering of transactions from the same sender, it's possible for the dapp to submit the quickminting transaction immediately after the deposit transaction, without having to wait for the deposit to be mined first.

Reply:
Ohhhhhhh, I think I get it.

When you deposit, you give the depositor a NFT representing their deposit.

The depositor can exchange their NFT for TBTC from the quickminter's treasury (if it has a large enough balance).

When we get around to sweeping the deposits, we mint TBTC to the quickminter's treasury.

The extra layer of indirection means that Mallory is only able to make unsweepable deposits (via DoS) to Wallet B and redeem from *the treasury* instead of just getting minted TBTC straight up. If the treasury has 30 TBTC, she could get all of that, but not everything.

> Because ethereum nonces enforce ordering of transactions from the same sender, it's possible for the dapp to submit the quickminting transaction immediately after the deposit transaction, without having to wait for the deposit to be mined first.

Can you elaborate here?

Reply:
Quickminter is an interesting concept but I think it requires two transactions: first one minting NFT, and second one proving the deposit has been swept and minting TBTC. There is also a question of how is the quickminter recharged - is it happening automatically with each successful sweep and if so, who is sending the transaction? Or maybe it is recharged manually by DAO and in this case, how often should it happen, and who should pay for it?

I think it could be worth taking a step back and agree on what we are trying to achieve by sweeping in batches instead of sweeping individually. I think the most important goal is to reduce the fees on Bitcoin transactions and the fact user has time to refund is just a nice side effect.

An alternative could be to let the user decide what is their refund timeout and allow users to claim their TBTC only when the deposit has been swept:
- Eve deposits 1 BTC and sets the refund time to T,
- Eve reveals her Ethereum address and blinding factor off-chain,
- When T comes, the wallet sweeps Eve's UTXO plus other pending UTXOs with similar T
- Eve proves the sweep happened on-chain and receives TBTC.

I think in this model, the amount of TBTC Eve receives could depend on the Bitcoin transaction fee and the dApp could suggest selecting the time T to reduce this cost. If Eve agrees to wait so that her sweep can be batched, she would cut less from her BTC on sweeping fees.

Disclaimer: I have not validated this yet, it's still just a loose idea.

Reply:
> Quickminter is an interesting concept but I think it requires two transactions: first one minting NFT, and second one proving the deposit has been swept and minting TBTC. There is also a question of how is the quickminter recharged - is it happening automatically with each successful sweep and if so, who is sending the transaction? Or maybe it is recharged manually by DAO and in this case, how often should it happen, and who should pay for it?

My prototype mints the TBTC for the current holder of the UTXO NFT automatically when the wallet sweeps (the wallet submits the second transaction proving that the UTXO has been swept), so the quickminter effectively buys the UTXO NFT from the depositor and receives the minted TBTC when the sweep is performed.

I haven't been able to figure out a way to do it securely without proving the sweep on ethereum so I've been taking the necessity of this second transaction for granted. That's also why I wanted to pay the operators a lump sum of rewards when the sweep transaction is submitted (with the submitter once again getting some extra).

Reply:
> The extra layer of indirection means that Mallory is only able to make unsweepable deposits (via DoS) to Wallet B and redeem from the treasury instead of just getting minted TBTC straight up. If the treasury has 30 TBTC, she could get all of that, but not everything, because she can't keep looping after the treasury runs out.

Yes. That is a critical security feature as it lets us limit our exposure to this attack 

> Can you elaborate here?

For the quickminting, we need 2 transactions: the first one proving the BTC deposit and creating the UTXO NFT, and the second one selling the UTXO to the quickminter. The second transaction needs to know the identifier of the first transaction, and it needs to be executed after the first one.

If we assign each UTXO NFT a GUID that is predictable in advance from a view function, the dapp is able to create both transactions without waiting for the first one to be mined first. The second transaction will simply use the ID received from the view function.

Because ethereum transactions have nonces, it is possible to submit both transactions immediately and the second transaction cannot be mined before the first one.

This gives better UX as the user doesn't need to return later, but can complete the entire operation in a single click.

Reply:
I'm sold - I'll add this to the RFC

Reply:
Had a chat w/ Matt and Piotr Monday about this. One of the main concerns we want to solve earlier than later is making sure that we build in the freedom to create financial services other than straight minting from a deposit.

For instance, we would like to be able to deposit bitcoin and then use that as collateral to take out TBTC as a loan, rather than selling BTC directly for TBTC. In order to make this work, we'd like to have account balances instead of NFTs that represent UTXO's.